revscoring/features/wikitext/datasources/parsed.py

```diff
@@ -0,0 +1,239 @@
+        if name is None:
+            name = "{0}({1})".format(self._name + ".heading_titles_matching",
+                                     regex.pattern)
+        return filters.regex_matching(regex, self.heading_titles, name=name)
+
+    def headings_by_level(self, level, name=None):
+        """
+        Constructs a :class:`revscoring.Datasource` that generates a `list` of
+        all headers of a level.
+        """
+        if name is None:
+            name = "{0}({1})".format(self._name + ".headings_by_level",
+                                     level)
+        return filters.filter(HeadingOfLevel(level).filter, self.headings,
+                              name=name)
+
+    def external_link_urls_matching(self, regex, name=None):
+        """
+        Constructs a :class:`revscoring.Datasource` that generates a `list` of
+        external link URLs that match a regular expression
+        """
+        if not hasattr(regex, "pattern"):
+            regex = re.compile(regex, re.I)
+
+        if name is None:
+            name = "{0}({1})" \
+                   .format(self._name + ".external_link_urls_matching",
+                           regex.pattern)
+
+        return filters.regex_matching(regex, self.external_link_urls,
+                                      name=name)
+
+    def wikilink_titles_matching(self, regex, name=None):
+        """
+        Constructs a :class:`revscoring.Datasource` that generates a `list`
+        of internal link titles names that match a regular expression.
+        """
+        if not hasattr(regex, "pattern"):
+            regex = re.compile(regex, re.I)
+
+        if name is None:
+            name = "{0}({1})" \
+                   .format(self._name + ".wikilink_titles_matching",
+                           regex.pattern)
+
+        return filters.regex_matching(regex, self.wikilink_titles, name=name)
+
+    def tag_names_matching(self, regex, name=None):
+        """
+        Constructs a :class:`revscoring.Datasource` that returns all tag names
+        that match a regular expression.
+        """
+        if not hasattr(regex, "pattern"):
+            regex = re.compile(regex, re.I)
+
+        if name is None:
+            name = "{0}({1})" \
+                   .format(self._name + ".tag_names_matching", regex.pattern)
+
+        return filters.regex_matching(regex, self.tag_names, name=name)
+
+    def template_names_matching(self, regex, name=None):
+        """
+        Constructs a :class:`revscoring.Datasource` that returns all template
+        names that match a regular expression.
+        """
+        if not hasattr(regex, "pattern"):
+            regex = re.compile(regex, re.I)
+
+        if name is None:
+            name = "{0}({1})" \
+                   .format(self._name + ".template_names_matching",
+                           regex.pattern)
+
+        return filters.regex_matching(regex, self.template_names, name=name)
+
+
+def _process_wikicode(text):
```

Comment:
These stuff (until line 218) I think it's better to be used as static methods in Revision class, we can use it without instantiating the REvision class, like:
from datasources import parsed
link = parsed.Revision._extract_external_link_url(foo)